derive-macros/src/lib.rs

```diff
@@ -134,3 +135,69 @@ fn gen_schema_fields(data: &Data) -> TokenStream {
     });
     quote! { #(#schema_fields),* }
 }
+
+/// Mark items as `internal_api` to make them public iff the `internal-api` feature is enabled.
+/// Note this doesn't work for inline module definitions (see `internal_mod!` macro in delta_kernel
+/// crate - can't export macro_rules! from proc macro crate).
+/// Ref: <https://github.com/rust-lang/rust/issues/54727>
+#[proc_macro_attribute]
+pub fn internal_api(
+    _attr: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let input = parse_macro_input!(item as Item);
+
+    // Create a version with public visibility for the unstable feature
+    let public_version = make_public(input.clone());
+
+    // The original item stays as-is for the non-unstable case
+    let output = quote! {
+        #[cfg(feature = "internal-api")]
+        #public_version
+
+        #[cfg(not(feature = "internal-api"))]
+        #input
+    };
+
+    output.into()
+}
+
+fn make_public(mut item: Item) -> Item {
+    fn set_pub(vis: &mut Visibility) -> Result<(), syn::Error> {
+        if matches!(vis, Visibility::Public(_)) {
+            Err(Error::new(
+                vis.span(),
+                "ineligible for #[internal_api]: item is already public",
+            ))
+        } else {
+            *vis = syn::parse_quote!(pub);
+            Ok(())
+        }
```

Comment:
tiny nit to consider?
```suggestion
        if matches!(vis, Visibility::Public(_)) {
            return Err(Error::new(
                vis.span(),
                "ineligible for #[internal_api]: item is already public",
            ))
        }
        *vis = syn::parse_quote!(pub);
        Ok(())
```